### PR TITLE
TREC DL and LLM AggreFact experiments for relevance benchmark + prompts comparisons and groundedness vs Bespoke Minicheck 7B

### PR DESCRIPTION
**Pull request by sfc-gh-dhuang**
# Description
Contain 2 main notebooks

1. `trec_dl_passage_relevance_judges.ipynb`: Experiments code with TREC DL passage level agreement and comparisons with Cortex team's hand-crafted prompt, UMBRELA prompt, and categorical prompt for search relevance LLM judges.

Doc with details: https://docs.google.com/document/d/1lwNC11S3LnNHglDNzcJNgPsd7zpOGNwTNktn8pI3bUc/edit?tab=t.nnid3pbdfr33

2. `llm_aggrefact_groundedness_benchmark.ipynb`: Experiments code with LLM-AggreFact benchmark for fact-checking grounding documents and comparing Bespoke Minicheck-7B vs trulens groundedness (gpt-4o). 
Doc: https://docs.google.com/document/d/1oMWDlk6JK6HP3IJQ_pjqB5KlHMtDHUKNoi5uTyWEYhk/edit?disco=AAABZ2Ar0vc 
## Other details good to know for developers

Please include any other details of this change useful for _TruLens_ developers.

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to
  not work as expected)
- [ ] New Tests
- [ ] This change includes re-generated golden test results
- [ ] This change requires a documentation update

----

> [!IMPORTANT]
> Add functions for generating balanced datasets and TREC DL passage benchmarks, and enhance feedback handling to support additional data types.
> 
>   - **New Functions**:
>     - `generate_balanced_llm_aggrefact_benchmark` and `generate_trec_dl_passage_benchmark` in `dataset_preprocessing.py` for creating balanced datasets and TREC DL passage benchmarks.
>   - **Feedback Handling**:
>     - Modified `run()` in `feedback.py` to support `dict` as a valid feedback function output type.
>     - Changed return value in `run_score_generation_on_single_row()` in `tru_benchmark_experiment.py` to `float('nan')` for invalid results.
>   - **Misc**:
>     - Added imports for `defaultdict` and `ir_datasets` in `dataset_preprocessing.py`.
> 
> <sup>This description was created by </sup>[Ellipsis](https://www.ellipsis.dev?ref=truera%2Ftrulens&utm_source=github&utm_medium=referral)<sup> for 35432e79a0bfc293b5c1e50b37ffa95cb6903057. It will automatically update as commits are pushed.</sup>

### CODE REVIEW

File: src/benchmark/trulens/benchmark/benchmark_frameworks/experiments/context_relevance/trec_dl_passage_relevance_judges.ipynb [hunk 1]
```diff
@@ -0,0 +1,1017 @@
+{
```
Comment (sfc-gh-jreini) [inline src/benchmark/trulens/benchmark/benchmark_frameworks/experiments/context_relevance/trec_dl_passage_relevance_judges.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #7.    def scoreddocs_qrels_confusion_matrix(
</pre><p>Can you break this function up? A lot going on here, hard to follow</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/truera/trulens/pull/1660/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

Comment (sfc-gh-jreini) [inline src/benchmark/trulens/benchmark/benchmark_frameworks/experiments/context_relevance/trec_dl_passage_relevance_judges.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #21.    trec_combined_df.to_csv(
</pre><p>do you need to write to csv in this notebook? suggesting to remove</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/truera/trulens/pull/1660/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='6'/>

Comment (sfc-gh-jreini) [inline src/benchmark/trulens/benchmark/benchmark_frameworks/experiments/context_relevance/trec_dl_passage_relevance_judges.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #42.    old_criteria = """
</pre><p>this is current criteria right? can you rename variable</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/truera/trulens/pull/1660/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='10'/>

Comment (sfc-gh-jreini) [inline src/benchmark/trulens/benchmark/benchmark_frameworks/experiments/context_relevance/trec_dl_passage_relevance_judges.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #8.    class CustomTermFeedback(Provider):
</pre><p>Should we have OOB feedbacks for these metrics?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/truera/trulens/pull/1660/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='12'/>

Comment (sfc-gh-jreini) [inline src/benchmark/trulens/benchmark/benchmark_frameworks/experiments/context_relevance/trec_dl_passage_relevance_judges.ipynb:1]: <p>Consider taking Gaurav's name off this prompt. Just call it prompt with fewshot examples or something</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/truera/trulens/pull/1660/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='15'/>

Comment (sfc-gh-dhuang) [reply src/benchmark/trulens/benchmark/benchmark_frameworks/experiments/context_relevance/trec_dl_passage_relevance_judges.ipynb:1]: removed

Comment (sfc-gh-dhuang) [reply src/benchmark/trulens/benchmark/benchmark_frameworks/experiments/context_relevance/trec_dl_passage_relevance_judges.ipynb:1]: thanks for catching this, updated and removed

Comment (sfc-gh-dhuang) [reply src/benchmark/trulens/benchmark/benchmark_frameworks/experiments/context_relevance/trec_dl_passage_relevance_judges.ipynb:1]: renamed to current_criteria

**Review by sfc-gh-jreini — APPROVED** [review]
